File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   val: number
 *   left: TreeNode | null
 *   right: TreeNode | null
 *   constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 *   }
 * }
 */

function rightSideView(root: TreeNode | null): number[] {
  const levels: TreeNode[][] = root ? [[root]] : [[]]
  while (levels.at(-1).length) {
    levels.push([])
    const next: TreeNode[] = levels.at(-1)
    levels.at(-2).forEach(node => {
      if (node.left) next.push(node.left)
      if (node.right) next.push(node.right)
    })
  }
  if (!levels.at(-1).length) levels.pop()
  return levels.map(level => level.at(-1).val)
}

File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.rb
# Definition for a binary tree node.
# class TreeNode
#   attr_accessor :val, :left, :right
#   def initialize(val = 0, left = nil, right = nil)
#     @val = val
#     @left = left
#     @right = right
#   end
# end
# @param {TreeNode} root
# @return {Integer[]}
def right_side_view(root)
  levels = root ? [[root]] : [[]]
  while !levels[-1].empty?
    levels << []
    next_level = levels[-1]
    levels[-2].each do |node|
      next_level << node.left if node.left
      next_level << node.right if node.right
    end
  end
  levels.pop if levels[-1].empty?
  levels.map{ |level| level[-1].val }
end